lib/logger.js

```diff
@@ -1,456 +1,634 @@
-  } else {
-    this._flushLimit = configs.FLUSH_BYTE_LIMIT
-  }
+    if (has(options, 'flushLimit')) {
+      if (!Number.isInteger(options.flushLimit)) {
+        const err = new TypeError('flushLimit must be an integer')
+        err.meta = {got: options.flushLimit}
+        throw err
+      }
+      this.flushLimit = options.flushLimit
+    }
 
-  if (options.retryTimes && Number.isInteger(options.retryTimes)) {
-    this._retryTimes = options.retryTimes
-  } else {
-    this._retryTimes = configs.RETRY_TIMES
-  }
+    if (has(options, 'flushInterval')) {
+      if (!Number.isInteger(options.flushInterval)) {
+        const err = new TypeError('flushInterval must be an integer')
+        err.meta = {got: options.flushInterval}
+        throw err
+      }
+      this.flushInterval = options.flushInterval
+    }
 
-  if (options.flushInterval && Number.isInteger(options.flushInterval)) {
-    this._flushInterval = options.flushInterval
-  } else {
-    this._flushInterval = configs.FLUSH_INTERVAL
-  }
+    if (has(options, 'baseBackoffMillis')) {
+      if (!Number.isInteger(options.baseBackoffMillis)
+        || options.baseBackoffMillis <= 0) {
 
-  if (options.retryTimeout && Number.isInteger(options.retryTimeout)) {
-    this._retryTimeout = options.retryTimeout
-  } else {
-    this._retryTimeout = configs.BACKOFF_PERIOD
-  }
+        const err = new TypeError('baseBackoffMillis must be an integer > 0')
+        err.meta = {got: options.baseBackoffMillis}
+        throw err
+      }
+      this.baseBackoffMillis = options.baseBackoffMillis
+    }
 
-  if (options.failedBufRetentionLimit && Number.isInteger(options.failedBufRetentionLimit)) {
-    this._failedBufRetentionLimit = options.failedBufRetentionLimit
-  } else {
-    this._failedBufRetentionLimit = configs.FAILED_BUF_RETENTION_LIMIT
-  }
+    if (has(options, 'maxBackoffMillis')) {
+      if (!Number.isInteger(options.maxBackoffMillis)
+        || options.maxBackoffMillis <= 0
+        || options.maxBackoffMillis < this.baseBackoffMillis) {
+
+        const err = new TypeError(
+          'maxBackoffMillis must be an integer > 0 and > baseBackoffMillis'
+        )
+        err.meta = {
+          got: options.maxBackoffMillis
+        , baseBackoffMillis: this.baseBackoffMillis
+        }
+        throw err
+      }
+      this.maxBackoffMillis = options.maxBackoffMillis
+    }
 
-  const {shimProperties} = options
-  if (shimProperties && shimProperties instanceof Array && shimProperties.length) {
-    this._shimProperties = options.shimProperties
-  }
+    if (has(options, 'shimProperties')) {
+      const val = options.shimProperties
+      if (!Array.isArray(val) || !val.length) {
+        const err = new TypeError('shimProperties must be a non-empty array')
+        err.meta = {got: val}
+        throw err
+      }
+      this.shimProperties = val
+    }
+
+    if (has(options, 'max_length')) {
+      const err = new Error('Deprecated.  max_length is no longer an option')
```

Comment:
if we're throwing, is this really _deprecated_? i could go either way on whether throwing vs console warning+ignoring is the right approach here, more of a messaging question.

should we indicate what the correct option is? i didn't see that mentioned elsewhere in docs but may be missing it.

Reply:
There is no correct option any more, which is why it doesn't say "deprecated."  We should avoid swallowing the error condition, which is why it throws.  If users think that their `max_length` is being honored, that'd be a problem.  And I don't think a `console.warn` would necessarily be seen...

Reply:
Perhaps just `Removed.  ...`?  I agree, the word choice implies that it works like the other warnings, but it doesn't.

Reply:
yeah just the wording, `Removed. ...` seems more appropriate. as a user I probably wouldn't expect an error to be thrown labeled as a deprecation, since deprecation implies it still works (but is discouraged).

Reply:
fixed